common/src/sql/dbinit.sql

```diff
@@ -128,6 +139,29 @@ CREATE INDEX ON omicron.public.service (
     sled_id
 );
 
+/*
+ * A table describing resource usage which may be associated
+ * with a collection of objects, including:
+ * - Projects
+ * - Organizations
+ * - Silos
+ * - Fleet
+ */
+CREATE TABLE omicron.public.resource_usage (
```

Comment:
Would you prefer `resource_allocation`? Or `collection_resource_allocation`?

Reply:
This goes even farther, and in conjunction with Dave's comment above, I found it confusing to see a `ResourceUsage` type that in fact doesn't have anything to do with resource usage but rather of what RFD 70 calls "provisioned" and is used here in a similar way.  Resource usage is not something that can be obtained from a database other than as a discrete (and therefore incomplete) historical record of past use: the answer to the question "how many pages of RAM provisioned for instance X are currently dirty?" is known if at all only to the kernel on the sled running that instance (if it's running at all).  So what this is really providing is a history of changes to virtual resource provisioning.  That's fine and useful but it's not resource usage.

The terminology in RFD 70 isn't necessarily as complete or sharp as we might like; we should improve or otherwise alter it to suit our needs, then use it religiously everywhere.  In my notes here, I'll declare the following taxa: a kind refers to a collection of similar resources even if they are not entirely fungible (e.g., virtual pages of RAM, physical CPU cores), a disposition describes the actions that may be taken on a defined subset of one kind of resource (e.g., a used page of RAM cannot have any action taken against it until, at a minimum, the instance using it ceases to do so; a provisioned extent of physical storage may be used by the instance to which it is provisioned or it may be deprovisioned by an administrative user), and a scope describes the collection of resources over which aggregated data are collected, presented, or used internally (e.g., what are called collections here but scopes in RFD 304 -- instance, project, silo -- or in principle any of what RFD 24 calls coherency domains).  My choice of taxa is merely for convenience; I made it rather arbitrarily for the sole purpose of being able to comment coherently, and it should not be taken as a recommendation for RFD 70 or otherwise.  There is also some sloppy terminology in RFD 312 that may be more distraction than real problem: for example, storage allocations are intended to be made with "fleet-wide context", yet in the RFD 24 sense there is never any use for context at a scope larger than the availability zone when making storage scheduling decisions because storage is not intended to be shared across this boundary.  I point this out not to be mean but as a reminder of just how important it is to be crisp with our terminology lest we allow it to lead us to incorrect architectural choices.  It's also important to note that we might find it useful to track and present resource data across scopes that exceed the domain over which the resource may be scheduled, but it is unlikely that we will need or be able to use those data for internal purposes like scheduling decisions.  This may or may not offer opportunities to simplify or otherwise optimise our software but does highlight the value of ascribing either informative or normative purposes to these data; those maintained solely for informative purposes may have less stringent implementation requirements.

Some of the rest of my thoughts here will go into a reply to Dave's other note and are really more suited to an RFD (304, maybe, though it's not quite discussing the same thing) than a code review -- I did read chunks of the code, but I am way too far removed from this work to have really useful comments at that level -- but one other thing I would offer a strong caution on here is that while it may be convenient to apply the same terms (used, provisioned, reserved, etc.) to resources of different kinds and to present historical data about each kind of resource in a uniform way, these resource kinds behave quite differently from one another on several different axes.  When we're providing a consumer with information about virtual resource provisioning, it may make sense to do as you've done here; these are simply numbers changed through administrative actions that result in software running inside the instance seeing a different quantity of that kind of resource, so any kind of provisioned virtual resource could be represented in the same manner.  This isn't the case for physical resources, nor is it necessarily the case for some dispositions of certain kinds of virtual resources.  As an example that brings this together, there is no such thing as capacity when talking about virtual disk space, even though there most certainly is at any given time some maximum amount of additional virtual disk space that the control plane may allow to be provisioned.  But if we assume, as RFD 70 does and as I vehemently advise we do always and everywhere, that there is no overprovisioning of any physical resource, then there is a very well defined capacity figure for virtual RAM pages at any given moment in time: it's the same as the number of physical RAM pages present over the scope of aggregation less those reserved for non-instance uses.  What I am suggesting, then, is that if we consider the 3D matrix with axes resource kind, physical or virtual, and disposition (whether with a fourth axis for scope in a denormalised table/s or not), the right schema may not be the one in which all matrix coordinate tuples are reduced to columns in a single otherwise uniform table (here, `omicron.public.resource_usage` below).  Indeed, this seems somewhat at odds with the unrolling concept described by RFD 161 for what appears to my naive eye to be a similar use case.  It is possible that we may want to present, and may need to record for our own consumption, sets of attributes for each entry in this matrix that are at most partially overlapping.  Some entries may not be meaningful, and the ideal semantics of the data may vary from one entry to another.  At minimum, it would seem to behoove us to consider this slightly more carefully before proceeding, or at least to do so in parallel with the expectation that there will be invalidating changes between now and RR.

Reply:
## On this PR...

Before responding to any specifics, I do want to call-out the intent of this PR: to surface collection-layer information about resources, such that they can be viewed by an operator. This PR is not sufficient to set capacities, it doesn't enforce anything, it also doesn't know about physical utilization -- it's simply trying to output info that Nexus knows to be true.

## On naming ...

Agreed that this should be updated. The name `resource_usage` itself is definitely too broad, and this table really should be limited to "Nexus-controlled accounting information for **collections** (or scopes, as you called them)". In the parlance of RFD 70, I could see either **provisioning information** (such as "how many CPUs are provisioned for instances across this project") or **reservation information**, if we can determine how to apply that at a scope-level. However, given that we're merely reporting what we've provisioned virtually, perhaps it's best to leave it at that.

I may call this `CollectionVirtualResourceProvisioning` for now - it's a bit verbose, but I believe it's accurate. We are currently expressing **user collection-level** information about what **virtual resources** have been **provisioned** by Nexus - not necessarily what's "in-use" right now, nor any info about physical manifestation of those resources.

## On capacity...

Definitely agreed that an RFD is necessary here. One of the key pieces that this PR does *not* do is track reservation usage of physical resources - for example, if a particular U.2 device has 1 TiB of space, we reserve 100 GiB for internal services, and we reserve 900 GiB for "disk storage", we *should* be tracking that reservation info in Nexus, so we know where to allocate regions backing virtual disks. This will be necessary to avoid over-committing, and Nexus will also likely need to keep track of reservations, not only for metrics, but for correctness.

As an idea, we could have a different table, perhaps called `PhysicalDiskReservations`, which tracks usage of reserved disk space at the scope of "zpool", "sled", "rack", and "az". However, the question of whether or not to denormalize in each use-case shouldn't be taken lightly, so admittedly, in a fully normalized world, we'd only need to actually track usage by individual physical disks.

Additionally, separate from virtual provisioning and physical reservations, we should be tracking actual utilization information reported by ZFS. This likely does not need a table in Cockroach, and can be emitted directly to Clickhouse, since it's not Nexus-controlled data, nor a table denormalization of Nexus-controlled data.

I do think there's an open question of "whether or not we want to even *try* to map the virtual usage to physical usage" - for example, if we discover that our the physical reservation of disk space has met capacity limits, should we bother trying to identify a silo/org/project that might be culpable? Or should we assert that "physical resources cannot be mapped to virtual resources", and let the operator figure out what to do?

## On table uniformity...

100% agree that different data for usage comes from different sources; one table does not need to fulfill all these dimensions. As mentioned earlier, I think tracking physical resource reservations will exist in separate tables, and utilization information likely won't be stored in Nexus at all (just reported directly to Clickhouse).

Reply:
Names updated in https://github.com/oxidecomputer/omicron/pull/1782/commits/c69363082a87e125f0fa743c1ecfb3949753ad10

Reply:
I think you may want to search for other references to "usage" in this PR because there still seem to be a lot of them where it means to say "reservation" or something else.

Reply:
I've tried taking on a pass of "usage", updating it to "provision" throughout this PR